File: java/CloneGraph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;

    public Node() {}

    public Node(int _val,List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
class Solution {
    public Node cloneGraph(Node node) {
        Map<Node, Node> map = new HashMap<>();
        Queue<Node> queue = new LinkedList<>();
        queue.offer(node);
        
        while (!queue.isEmpty()) {
            Node topNode = queue.poll();
            
            if (!map.containsKey(topNode)) {
                Node copy = new Node(topNode.val, new ArrayList<>());
                map.put(topNode, copy);
            }
            
            Node copy = map.get(topNode);
            List<Node> neighbors = topNode.neighbors;
            
            for (Node n : neighbors) {
                if (!map.containsKey(n)) {
                    Node c = new Node(n.val, new ArrayList<>());
                    map.put(n, c);
                    queue.offer(n);
                }
                
                copy.neighbors.add(map.get(n));
            }
        }
        
        return map.get(node);
    }
}

File: archive/python/bfs/CloneGraph.py
# Definition for a undirected graph node
class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if node is None:
            return 
        
        d = {}
        queue = [node]
        new_node = UndirectedGraphNode(node.label)
        d[node] = new_node
        while len(queue) > 0:
            curr = queue.pop(0)
            for i in xrange(len(curr.neighbors)):
                if curr.neighbors[i] not in d:
                    copy = UndirectedGraphNode(curr.neighbors[i].label)
                    d[curr].neighbors.append(copy)
                    d[curr.neighbors[i]] = copy
                    queue.append(curr.neighbors[i])
                else:
                    d[curr].neighbors.append(d[curr.neighbors[i]])
        
        return new_node

